# 

![](img/screenshot-bmcgenomics.biomedcentral.com-2021.03.15-16_04_03.png){height=70mm}

\note{\tiny\begin{itemize}
\item in this session i'm going to talk about a genome assembly from 2019 that took a long read + scaffolding approach to re-assemble a genome that had been assembled with older technology and had some issues
\item i like this example because they really tried all the technologies they could get their hands on and then assessed the effect of each one on the assembly
\item it shows you how there isn't necessarily one approach that's going to work for every genome
\end{itemize}
}

# *Apis mellifera*: Western honey bee

![](img/Apis_mellifera_-_Brassica_napus_-_Valingu.jpg){height=70mm}

\source{
   \href{https://commons.wikimedia.org/wiki/File:Apis_mellifera_-_Brassica_napus_-_Valingu.jpg}{ Ivar Leidus via Wikimedia Commons}
}

\note{\tiny\begin{itemize}
\item why the honeybee?
\item honeybees are agriculturally quite important
\item obviously they are used commercially to produce honey
\item they're also important in pollination services to other agricultural sectors
\item despite their commercial importance they haven't really been through an extensive selection process like other livestock such as sheep and cattle
\item there are now genetic improvement programs in place in australia and new zealand to address this
\item these programs combine genome-wide genotyping of honeybees with measurements of hive performance to find markers that could be associated with performance
\item this is livestock improvement process called genomic selection
\item plus, they are cute so why wouldn't you want to work on them
\item they are also pretty interesting from an evolutionary point of view because of their unusual social structure and sex determination system
\end{itemize}
}

# First genome assembly: Amel_4.0

![](img/41586_2006_Article_BFnature05260_Fig2_HTML.jpg){height=70mm}

\source{
   \href{https://doi.org/10.1038/nature05260}{Honeybee Genome Sequencing Consortium. Nature, 2006}
}

\note{\tiny\begin{itemize}
\item the first honeybee genome was released in 2006
\item it's called the Amel\_4.0 assembly for apis mellifera
\item that's an internal name from the lab doing the assembly, it doesn't mean it was the fourth genome released
\item this was a shotgun plus sanger sequencing approach
\item like the one i described for the first human genome assembly in the lecture material
\item they managed to assemble scaffolds for all of the 16 honeybee chromosomes, but there were lots of gaps caused by issues with coverage of certain regions, and the number of genes was quite low
\end{itemize}
}

# 

![](img/screenshot-onlinelibrary.wiley.com-2021.03.16-11_09_47.png){height=70mm}

\note{\tiny\begin{itemize}
\item it was quite a surprise to find such a low number of genes
\item for a while we thought that this was actually what the honeybee genome looked like
\item the theory in this paper was that one of the reasons honeybee colonies were so sensitive and could collapse without warning was that they simply didn't have the reservoir of genes to detoxify pesticides that were being used on crops near their hives
\end{itemize}
}

# First major upgrade: Amel_4.5

![](img/screenshot-bmcgenomics.biomedcentral.com-2021.03.15-16_12_18.png){height=70mm}

\note{\tiny\begin{itemize}
\item the first major update to this genome came in 2016, using some short read DNA and RNA sequencing improved the number of genes and filled in some gaps
\item the assembly was still quite fragmented with gaps in repetitive regions
\item the point here is not that the people who did these first two assemblies did something wrong, or are bad scientists
\item far from it - the initial assemblies were a lot of work and very useful as reference genomes
\item the quality of the assemblies reflects the technology that was available at the time they were done
\item the point is to highlight how those technologies from even just a few years ago affected the quality of assemblies 
\item it's a pretty good time right now to be doing genome assembly because the technology has really come a long way in a short time
\end{itemize}
}

# Hybrid approach: Amel_HAv3

![](img/screenshot-bmcgenomics.biomedcentral.com-2021.03.15-16_04_03.png){height=70mm}

\note{\tiny\begin{itemize}
\item in this 2019 assembly they took advantage of that an used a hybrid approach
\item and this did produce a better assembly, so this is now the reference honeybee genome
\end{itemize}
}

# Kitchen sink genome assembly

![](img/12864_2019_5642_MOESM6_ESM.pdf){height=70mm}

\note{\tiny\begin{itemize}
\setlength{\itemsep}{1pt}
\setlength{\parskip}{0pt}
\setlength{\parsep}{0pt}
\item they had a lot of data for this assembly
\item the blue boxes are the data sources
\item in the lecture material i talked about linkage maps, pacbio (long reads), bionano (optical mapping) and hi-c (proximity ligation)
\item i didn't mention 10x
\item 10x is a type of short-read illumina sequencing where a bit of microfludics magic is used to tag read that come from the same genomic region
\item it just puts each input fragment of dna into a tiny bubble and then tags all the reads that are generated from that bubble with a barcode
\item it's still short-read sequencing, but instead of having to assemble the whole genome at once, you just assemble bits of it and then put all the bits together
\item the flowchart shows the steps they used to generate the final assembly, with the software names in yellow
\item step 1. generate contigs from the pacbio data
\item step 2. scaffold contigs with the 10x linked reads
\item step 3. oriented contigs and scaffolds with the genetic map, then used the long read to fill gaps (HAv1)
\item step 4. another round of scaffolding with optical mapping data (HAv2)
\item step 5. another round of scaffolding with the genetic map (again) and the proximity ligation data
\item step 6. another round of gap filling, then polishing (HAv3)
\end{itemize}
}

# Contiguity is much better

![](img/12864_2019_5642_Fig1_HTML.png){height=70mm}

\note{\tiny\begin{itemize}
\item this is weird diagram
\item the number of bars is proportional to the number of contigs or scaffolds
\item then the size of the scaffold is on the x-axis as usual
\item more contigs or scaffolds mean more fragmented, so a worse assembly
\item left is the upgraded amel\_4.5 assembly
\item they also show an assembly from the 10x alone, so you remember i said this was just illumina short reads, and you can see it doesn't really provide a much better assembly by itself
\item then pacbio contigs alone
\item then pacbio scaffolded with 10x, so the linked reads do help with the scaffolding
\item then the final hybrid assembly
\item the main thing to note is HOW MUCH BETTER the LR assembly is than the short read assembly alone
\item this is just because of tech advances that made long read sequencing increasingly available
\item wasn't available for the 2006 assembly
\end{itemize}
}

# 10\% increase in genome size, 21\% increase in repeat size

![](img/12864_2019_5642_Fig5_HTML.png){height=70mm}

\note{\tiny\begin{itemize}
\item this graph just shows the sequence they added compared to the previous assembly
\item dark blue is what was in Amel\_4.5
\item dark grey plus green is the sum of the new material in the hybrid assembly
\item and the categories are just different classes of repetitive sequence like transposons and satellite repeats
\item aluI and avaI are telomeres and centromeres respectively, you can see they weren't present at all before
\end{itemize}
}

# Some metrics

```{r metrics}
metrics <- fread("data/metrics_hav3.csv",
           skip = 1,
           select = 2:7,
           header = TRUE)

pander::pander(metrics,
               big.mark = ",",
               emphasize.strong.cols = 1)
```
\vspace{1ex}

\scriptsize\centering

 **1**. BUSCO, ***Benchmarking Universal Single-Copy Orthologs***  
 **2**. Illumina data from two individual honeybees from NZ

\note{\tiny\begin{itemize}
\item i've taken the three honeybee genomes and compared them with the metrics i talked about in video three
\item (scaffold n50) the big difference between the original assembly and the upgraded short-read assembly is the scaffolding
\item whereas the main advantage to the hybrid assembly is that the contigs are much better (higher contig N50)
\item that's because the contigs were assembled with pacbio long reads
\item these are the VGP target metrics again, and you can see the newer hybrid assembly meets their requirements
\item i couldn't measure kmer completeness because i didn't have an error-corrected set of illumina reads to do a kmer analysis
\item the "genes" metric comes from BUSCO, which is an algorithm that searches the genome assembly for certain genes from closely related species
\item the BUSCO scores are quite high for all three assemblies, but we can see the hybrid assembly is much better - shows that you can't rely on any one metric
\item the last row isn't transcript mappings, it's illumina sequencing data from two individual honeybees that are part of the genomic selection program in NZ
\item and those reads do map better to the hybrid assembly, suggesting it would be a better reference for genotyping NZ honeybees
\end{itemize}
}

# Is this a representative genome?

![](img/13592_2020_836_Fig2_HTML.png){height=70mm}

\source{
   \href{https://doi.org/10.1007/s13592-020-00836-4}{Carpenter \& Harpur, 2021}
}

\note{\tiny\begin{itemize}
\item so we have an improved reference, is that the end of the genome assembly story for honeybees?
\item as always, there's more to it than that
\item all three assemblies were done from honeybees from the US
\item we know honeybees were introduced to the east coast of the US from Europe in the late 1800s and early 1900s
\item they were introduced to Australia and NZ from Europe around the same time
\item so theoretically, they could have quite a different genetic history, and they've been isolated for maybe 100s of generations
\end{itemize}
}

# We can assemble local genomes to compare

![](fig/wgas-1.eps){height=70mm}

\note{\tiny\begin{itemize}
\item as i said, now is the time to be doing genome assemblies, so we can easily assemble our own genomes to compare
\item these are the two individuals from the genomic improvement program i mentioned
\item these honeybees have personalised genomes and they didn't even have to pay a biotech company for it
\item this is oxford nanopore long-read data
\item this probably cost around \$500 per genome, and it took half a day in the lab and half a day of bioinformatics
\item these are whole-genome alignments against the hybrid assembly to see if the local genome assemblies are ok, and they're reasonable - scaffold N50 around 8 Mb
\item the difference in these two assemblies is just down to the difference in the quality of the input DNA
\item i mentioned that was important in one of the other videos
\end{itemize}
}

# Genomic rearrangements

![](fig/inversions.eps){height=70mm}

\note{\tiny\begin{itemize}
\item and this is a closer look at the alignment of these two genomes against one of the reference chromosomes
\item both of these individuals have a 2.5 Mb inversion on this chromosome
\item so these honeybees are the same species as US honeybees
\item is the genome the same?
\item no. ordinary reference genomes are just a snapshot of the genomes in the individuals that were sequenced
\item variants are out there and they need to be incorporated into the reference
\end{itemize}
}

# Graphs and pangenomes

![](img/screenshot-genomebiology.biomedcentral.com-2021.03.16-14_41_29.png){height=30mm}

\vspace{2ex}

![](img/screenshot-genomebiology.biomedcentral.com-2021.03.16-14_41_49.png){height=30mm}

\note{\tiny\begin{itemize}
\item so the field is starting to move towards treating reference genomes as graphs that capture all sorts of variants, rather than a linear sequence representing the haploid genetic material of one individual
\end{itemize}
}
