# 

![](img/screenshot-bmcgenomics.biomedcentral.com-2021.03.15-16_04_03.png){height=70mm}


::: notes
\tiny
- in this session i'm going to talk about a genome assembly from 2019 that took a long read + scaffolding approach to re-assemble a genome that had been assembled with older technology and had some issues
- i like this example because they really tried all the technologies they could get their hands on and then assessed the effect of each one on the assembly
- it shows you how there isn't necessarily one approach that's going to work for every genome
:::

# *Apis mellifera*: Western honey bee

![](img/Apis_mellifera_-_Brassica_napus_-_Valingu.jpg){height=70mm}

\source{
   \href{https://commons.wikimedia.org/wiki/File:Apis_mellifera_-_Brassica_napus_-_Valingu.jpg}{ Ivar Leidus via Wikimedia Commons}
}

::: notes
\tiny
- why the honeybee?
- honeybees are agriculturally quite important
- used commercially to produce honey
- important in pollination services to other agricultural sectors
- despite their commercial importance haven't been through extensive selection process like other livestock (sheep and cattle)
- genetic improvement programs in australia and new zealand
- combine genome-wide genotyping + measurements of hive performance -> markers that could be associated with performance
- modern livestock improvement process called genomic selection
- also interesting from an evolutionary point of view because of their unusual social structure and interesting behaviours
:::

# First genome assembly: Amel_4.0

![](img/41586_2006_Article_BFnature05260_Fig2_HTML.jpg){height=70mm}

\source{
   \href{https://doi.org/10.1038/nature05260}{Honeybee Genome Sequencing Consortium. Nature, 2006}
}

::: notes
\tiny
- first honeybee genome released in 2006
- called the Amel\_4.0 assembly for apis mellifera
- 4.0 is an internal name from the lab doing the assembly (doesn't mean it was the fourth genome released)
- shotgun plus sanger sequencing approach
- like first human genome assembly in the lecture material
- scaffolds for all of the 16 honeybee chromosomes, but
    - lots of gaps caused by issues with coverage of certain regions
    - number of genes was quite low
:::

# 

![](img/screenshot-onlinelibrary.wiley.com-2021.03.16-11_09_47.png){height=70mm}

::: notes
\tiny
- surprise to find such a low number of genes
- for a while we thought that this was actually what the honeybee genome looked like
- the theory in this paper was that one of the reasons honeybee colonies were so sensitive and could collapse without warning was that they simply didn't have the reservoir of genes to detoxify pesticides that were being used on crops near their hives
:::

# First major upgrade: Amel_4.5

![](img/screenshot-bmcgenomics.biomedcentral.com-2021.03.15-16_12_18.png){height=70mm}

::: notes
\tiny
- the first major update to this genome came in 2016, using some short read DNA and RNA sequencing improved the number of genes and filled in some gaps
- the assembly was still quite fragmented with gaps in repetitive regions
- the point here is not that the people who did these first two assemblies did something wrong, or are bad scientists
- actually the initial assemblies were a lot of work and very useful as reference genomes
- the quality of the assemblies reflects the technology that was available at the time they were done
- the point is to highlight how those technologies from even just a few years ago affected the quality of assemblies 
- it's a pretty good time right now to be doing genome assembly because the technology has really come a long way in a short time
:::

# Hybrid approach: Amel_HAv3

![](img/screenshot-bmcgenomics.biomedcentral.com-2021.03.15-16_04_03.png){height=70mm}

::: notes
\tiny
- in this 2019 assembly they took advantage of that an used a hybrid approach
- and this did produce a better assembly, so this is now the reference honeybee genome
:::

# Kitchen sink genome assembly

![](img/12864_2019_5642_MOESM6_ESM.pdf){height=70mm}

::: notes
\tiny
- they had a lot of data for this assembly
- the blue boxes are the data sources
- in the lecture material i talked about linkage maps, pacbio (long reads), bionano (optical mapping) and hi-c (proximity ligation)
- i didn't mention 10x
- 10x is a type of short-read illumina sequencing where a bit of microfludics magic is used to tag read that come from the same genomic region
- it just puts each input fragment of dna into a tiny bubble and then tags all the reads that are generated from that bubble with a barcode
- it's still short-read sequencing, but instead of having to assemble the whole genome at once, you just assemble bits of it and then put all the bits together
- the flowchart shows the steps they used to generate the final assembly, with the software names in yellow
- step 1. generate contigs from the pacbio data
- step 2. scaffold contigs with the 10x linked reads
- step 3. oriented contigs and scaffolds with the genetic map, then used the long read to fill gaps (HAv1)
- step 4. another round of scaffolding with optical mapping data (HAv2)
- step 5. another round of scaffolding with the genetic map (again) and the proximity ligation data
- step 6. another round of gap filling, then polishing (HAv3)
:::

# Contiguity is much better

![](img/12864_2019_5642_Fig1_HTML.png){height=70mm}

::: notes
\tiny
- this is weird diagram
- the number of bars is proportional to the number of contigs or scaffolds
- then the size of the scaffold is on the x-axis as usual
- more contigs or scaffolds mean more fragmented, so a worse assembly
- left is the upgraded amel\_4.5 assembly
- they also show an assembly from the 10x alone, so you remember i said this was just illumina short reads, and you can see it doesn't really provide a much better assembly by itself
- then pacbio contigs alone
- then pacbio scaffolded with 10x, so the linked reads do help with the scaffolding
- then the final hybrid assembly
- the main thing to note is HOW MUCH BETTER the LR assembly is than the short read assembly alone
- this is just because of tech advances that made long read sequencing increasingly available
- wasn't available for the 2006 assembly
:::

# 10\% increase in genome size, 21\% increase in repeat size

![](img/12864_2019_5642_Fig5_HTML.png){height=70mm}

::: notes
\tiny
- SKIP: main addition was repetetive regions that hadn't been assembled properly in the short read genome
- this graph just shows the sequence they added compared to the previous assembly
- dark blue is what was in Amel\_4.5
- dark grey plus green is the sum of the new material in the hybrid assembly
- and the categories are just different classes of repetitive sequence like transposons and satellite repeats
- aluI and avaI are telomeres and centromeres respectively, you can see they weren't present at all before
:::

# Some metrics

```{r metrics}
metrics <- fread("data/metrics_hav3.csv",
           skip = 1,
           select = 2:7,
           header = TRUE)

pander::pander(metrics,
               big.mark = ",",
               emphasize.strong.cols = 1)
```
\vspace{1ex}

\scriptsize\centering

 **1**. BUSCO, ***Benchmarking Universal Single-Copy Orthologs***  
 **2**. Illumina data from two individual honeybees from NZ

::: notes
\tiny
- i've taken the three honeybee genomes and compared them with the metrics i talked about in the lecture
- (scaffold n50) the big difference between the original assembly and the upgraded short-read assembly is the scaffolding
- whereas the main advantage to the hybrid assembly is that the contigs are much better (higher contig N50)
- that's because the contigs were assembled with pacbio long reads
- these are the VGP target metrics again, and you can see the newer hybrid assembly meets their requirements
- i couldn't measure kmer completeness because i didn't have an error-corrected set of illumina reads to do a kmer analysis
- the "genes" metric comes from BUSCO, which is an algorithm that searches the genome assembly for certain genes from closely related species
- the BUSCO scores are quite high for all three assemblies, but we can see the hybrid assembly is much better - shows that you can't rely on any one metric
- the last row isn't transcript mappings, it's illumina sequencing data from two individual honeybees that are part of the genomic selection program in NZ
- and those reads do map better to the hybrid assembly, suggesting it would be a better reference for genotyping NZ honeybees
:::

# Is this a representative genome?

![](img/13592_2020_836_Fig2_HTML.png){height=70mm}

\source{
   \href{https://doi.org/10.1007/s13592-020-00836-4}{Carpenter \& Harpur, 2021}
}

::: notes
\tiny
- so we have an improved reference, is that the end of the genome assembly story for honeybees?
- all three assemblies were done from honeybees from the US
- we know honeybees were introduced to the east coast of the US from Europe in the late 1800s and early 1900s
- they were introduced to Australia and NZ from Europe around the same time
- so theoretically, they could have quite a different genetic history, and they've been isolated for maybe 100s of generations
- not sure if local honeybees have the same genome
:::

# We can assemble local genomes to compare

![](fig/wgas-1.eps){height=70mm}

::: notes
\tiny
- as i said, now is the time to be doing genome assemblies, so we can easily assemble our own genomes to compare
- these are the two individuals from the genomic improvement program i mentioned in nz
- this is oxford nanopore long-read data
- this probably cost around \$500 per genome, and it took half a day in the lab and half a day of bioinformatics
- these are whole-genome alignments against the hybrid assembly 
- these alignemnt plots connect blocks of homologous sequence between the reference and the query by a line
- these quick local genome assemblies are ok - scaffold N50 around 8 Mb
- the difference in these two assemblies is just down to the difference in the quality of the input DNA
- i mentioned that was important in one of the other videos
:::

# Genomic rearrangements

![](fig/inversions.eps){height=70mm}

::: notes
\tiny
- and this is a closer look zoomed in to one of the reference chromosomes
- both of these individuals have a 2.5 Mb inversion on this chromosome
- so these honeybees are the same species as US honeybees but their genomes are different
- ordinary reference genomes like the current honeybee reference genome are just a snapshot of the genomes in the individuals that were sequenced
- genome variants are out there and they need to be incorporated into the reference
:::

# Graphs and pangenomes

![](img/screenshot-genomebiology.biomedcentral.com-2021.03.16-14_41_29.png){height=30mm}

\vspace{2ex}

![](img/screenshot-genomebiology.biomedcentral.com-2021.03.16-14_41_49.png){height=30mm}

::: notes
\tiny
- the field is starting to move towards treating reference genomes as graphs that capture all sorts of variants, rather than a linear sequence representing the haploid genetic material of one individual
- you'll learn about advanced assembly topics like this later in the subject
:::